### PR TITLE
add claim, complete and heartbeat tasks to cli

### PR DESCRIPTION
**Pull request by TusharMohapatra07**
- This PR is linked to issue #436 . I've implemented the claim and complete task.
- Need some help in the following areas:
  1. found an endpoint for hearbeatTasks.go (https://github.com/resonatehq/resonate/blob/main/internal/app/coroutines/heartbeatTasks.go) but according to the documentation (https://resonatehq.github.io/resonate/#tag/Tasks/operation/completeTask), its a post request requiring just the process-id. Unable to find the endpoint to map id and counter with
  2. If this implementation is correct, I need to pass a type that implements `gocoro.Coroutine[*t_aio.Submission, *t_aio.Completion, any]`. In tasks.go i need to register these commands in order for them to work. I've so far figured out that gocoro.newCoroutine with signature `func[T, TNext, TReturn any](f CoroutineFunc[T, TNext, TReturn], r map[string]any) *coroutine[T, TNext, TReturn]` creates a coroutine that implements this type. But i'm unable to figure out what to pass in place of f and r

### CODE REVIEW

**Comment by dfarr** [comment]
Thank you @TusharMohapatra07!!!

The scaffolding here looks great, exactly what we need. But we do need to use the openapi auto generated client here, not the `Request`/`Response` objects which are internal facing structures for the server only. Here the cli is acting as a client and getting a response back from the server.

The promise commands and schedule commands should be good guides, for example let's take a look at [promise create](https://github.com/resonatehq/resonate/blob/main/cmd/promises/create.go).

1. Pass the client in to the command as a [function parameter](https://github.com/resonatehq/resonate/blob/main/cmd/promises/create.go#L21)
2. Take all the command line arguments and construct a params/body struct to pass to the [generated function](https://github.com/resonatehq/resonate/blob/main/cmd/promises/create.go#L69) (there will be functions corresponding to tasks too).
3. Grab the response and print a corresponding status message

You may find our api specifiction docs [here](https://resonatehq.github.io/resonate/#tag/Tasks/operation/claimTask) handy too. Let me know if I can help further!

**Comment by TusharMohapatra07** [comment]
@dfarr So, i need to use the functions present in github.com/resonatehq/resonate/pkg/client/v1. correct ?

**Comment by dfarr** [comment]
> @dfarr So, i need to use the functions present in github.com/resonatehq/resonate/pkg/client/v1. correct ?

Correct!

**Comment by TusharMohapatra07** [comment]
@dfarr on it.

**Comment by TusharMohapatra07** [comment]
@dfarr please review and suggest changes (if any).

File: cmd/tasks/claim.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+			params := &v1.ClaimTaskParams{
+				RequestId: &requestId, // Optional tracking ID
+			}
+
+			// Create the body for the claim task request
+			body := v1.ClaimTaskJSONRequestBody{
+				Id:        id,
+				Counter:   counter,
+				ProcessId: processId,
+				Ttl:       int64(ttl.Milliseconds()), // Convert duration to milliseconds
```
Comment (dfarr) [inline cmd/tasks/claim.go]: I believe `ttl.Milliseconds()` is already an int64 so we shouldn't have to convert it

File: cmd/tasks/claim.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+			res, err := c.V1().ClaimTaskWithResponse(context.TODO(), params, body)
+
+			if err != nil {
+				return err // Return any errors from the request
+			}
+
+			// Handle the response based on the status code
+			if res.StatusCode() == 201 {
+				cmd.Printf("Task claimed: %s\n", id)
+			} else if res.StatusCode() == 403 {
```
Comment (dfarr) [inline cmd/tasks/claim.go]: For consistency, I think we should [print the entire body](https://github.com/resonatehq/resonate/blob/main/cmd/promises/get.go#L37-L40) if the status code is not 201, this is what we do for promises and schedules and will keep our cli more flexible to changes. Could we follow the same pattern for all task commands?

**Review by dfarr — CHANGES_REQUESTED** [review]
This looks really good! I have a few small requests.

One additional thought: for `promises` and `schedules` the id parameter is a positional parameter (see [here](https://github.com/resonatehq/resonate/blob/main/cmd/promises/get.go#L22-L30)), I'm wondering if we should take a similar approach for tasks to be consistent, what do you think?

**Comment by TusharMohapatra07** [comment]
@dfarr I think having id as positional parameter consistently throughout the application enhances the user experience a lot. i'm implementing those changes you requested.

**Comment by TusharMohapatra07** [comment]
@dfarr please review. Also, the success code for claim and complete is 201 and that of heartbeat is 200 (according to docs). So, i've printed the response body when the status code is not equal to the above mentioned.

**Comment by dfarr** [comment]
Very good! Only one thing, the heartbeat request takes a single parameter of `--process-id`, whichever process id is specified on claim can be used to heartbeat, this way a client can claim a bunch of tasks and batch heartbeat with a single process id.

```
resonate tasks heartbeat --process-id foo
```

It must be hard to test this! Here is how I do it, maybe this will help
```
# start the server
go run ./... serve

# create a promise and a task
go run ./... promises create foo --timeout 24h --tag resonate:invoke=poll://default

# claim the task (the id is auto incrementing, starts with 1)
go run ./... tasks claim 1 --counter 1 --process-id foo --ttl 1m

# complete the task (if not completed in one minute, will need to be re-claimed, the counter will be bumped)
go run ./... tasks complete 1 --counter 1
```

If you have a sqlite db editor you can open up the db file (defaults to `resonate.db`) and inspect the tasks table to see how the cli command affects tasks.

One more nitpick, we don't need the request id for the cli. It doesn't hurt to keep it so I am of two minds on this one, but getting rid of it as a flag might make the cli a little simpler.

**Comment by TusharMohapatra07** [comment]
@dfarr I'll get rid of requestId. Sorry but i'm not clear about the first requirement. What should i do with the process-id ?   I'm not clear what changes i need to implement.

**Comment by TusharMohapatra07** [comment]
@dfarr please review

**Review by dfarr — COMMENTED** [review]
🎉

**Pull request merged**